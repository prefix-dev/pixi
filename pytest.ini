[pytest]
addopts = --basetemp=pytest-temp
tmp_path_retention_policy = failed
testpaths = tests
markers =
    slow: marks tests as slow (deselect with '-m "not slow"')
