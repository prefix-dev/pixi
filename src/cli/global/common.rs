use std::path::PathBuf;

use indexmap::IndexMap;
use miette::IntoDiagnostic;
use rattler_conda_types::{
    Channel, ChannelConfig, MatchSpec, PackageName, Platform, PrefixRecord, RepoDataRecord,
};
use rattler_repodata_gateway::sparse::SparseRepoData;
use rattler_solve::{resolvo, ChannelPriority, SolverImpl, SolverTask};
use reqwest_middleware::ClientWithMiddleware;

use crate::{
    config::{home_path, Config},
    prefix::Prefix,
    repodata,
    utils::reqwest::build_reqwest_clients,
};

/// Global binaries directory, default to `$HOME/.pixi/bin`
pub struct BinDir(pub PathBuf);

impl BinDir {
    /// Create the Binary Executable directory
    pub async fn create() -> miette::Result<Self> {
        let bin_dir = bin_dir().ok_or(miette::miette!(
            "could not determine global binary executable directory"
        ))?;
        tokio::fs::create_dir_all(&bin_dir)
            .await
            .into_diagnostic()?;
        Ok(Self(bin_dir))
    }

    /// Get the Binary Executable directory, erroring if it doesn't already exist.
    pub async fn from_existing() -> miette::Result<Self> {
        let bin_dir = bin_dir().ok_or(miette::miette!(
            "could not find global binary executable directory"
        ))?;
        if tokio::fs::try_exists(&bin_dir).await.into_diagnostic()? {
            Ok(Self(bin_dir))
        } else {
            Err(miette::miette!(
                "binary executable directory does not exist"
            ))
        }
    }
}

/// Global binary environments directory, default to `$HOME/.pixi/envs`
pub struct BinEnvDir(pub PathBuf);

impl BinEnvDir {
    /// Construct the path to the env directory for the binary package `package_name`.
    fn package_bin_env_dir(package_name: &PackageName) -> miette::Result<PathBuf> {
        Ok(bin_env_dir()
            .ok_or(miette::miette!(
                "could not find global binary environment directory"
            ))?
            .join(package_name.as_normalized()))
    }

    /// Get the Binary Environment directory, erroring if it doesn't already exist.
    pub async fn from_existing(package_name: &PackageName) -> miette::Result<Self> {
        let bin_env_dir = Self::package_bin_env_dir(package_name)?;
        if tokio::fs::try_exists(&bin_env_dir)
            .await
            .into_diagnostic()?
        {
            Ok(Self(bin_env_dir))
        } else {
            Err(miette::miette!(
                "could not find environment for package {}",
                package_name.as_source()
            ))
        }
    }

    /// Create the Binary Environment directory
    pub async fn create(package_name: &PackageName) -> miette::Result<Self> {
        let bin_env_dir = Self::package_bin_env_dir(package_name)?;
        tokio::fs::create_dir_all(&bin_env_dir)
            .await
            .into_diagnostic()?;
        Ok(Self(bin_env_dir))
    }
}

/// Global binaries directory, default to `$HOME/.pixi/bin`
///
/// # Returns
///
/// The global binaries directory
pub fn bin_dir() -> Option<PathBuf> {
    home_path().map(|path| path.join("bin"))
}

/// Global binary environments directory, default to `$HOME/.pixi/envs`
///
/// # Returns
///
/// The global binary environments directory
pub fn bin_env_dir() -> Option<PathBuf> {
    home_path().map(|path| path.join("envs"))
}

/// Returns the package name from a MatchSpec
///
/// # Returns
///
/// The package name from the given MatchSpec
fn package_name(package_matchspec: &MatchSpec) -> miette::Result<PackageName> {
    package_matchspec.name.clone().ok_or_else(|| {
        miette::miette!(
            "could not find package name in MatchSpec {}",
            package_matchspec
        )
    })
}

/// Get the friendly channel name of a [`PrefixRecord`]
///
/// # Returns
///
/// The friendly channel name of the given prefix record
pub(super) fn channel_name_from_prefix(
    prefix_package: &PrefixRecord,
    channel_config: &ChannelConfig,
) -> String {
    Channel::from_str(&prefix_package.repodata_record.channel, channel_config)
        .map(|ch| repodata::friendly_channel_name(&ch))
        .unwrap_or_else(|_| prefix_package.repodata_record.channel.clone())
}

/// Load package records from [`SparseRepoData`] for the given package MatchSpec
///
/// # Returns
///
/// The package records (with dependencies records) for the given package MatchSpec
pub fn load_package_records<'a>(
    package_matchspec: MatchSpec,
    sparse_repodata: impl IntoIterator<Item = &'a SparseRepoData>,
) -> miette::Result<Vec<RepoDataRecord>> {
    let package_name = package_name(&package_matchspec)?;
    let available_packages =
        SparseRepoData::load_records_recursive(sparse_repodata, vec![package_name], None)
            .into_diagnostic()?;
    let virtual_packages = rattler_virtual_packages::VirtualPackage::current()
        .into_diagnostic()?
        .iter()
        .cloned()
        .map(Into::into)
        .collect();

    // Solve for environment
    // Construct a solver task that we can start solving.
    let task = SolverTask {
        specs: vec![package_matchspec],
        available_packages: &available_packages,
        virtual_packages,
        locked_packages: vec![],
        pinned_packages: vec![],
        timeout: None,
        channel_priority: ChannelPriority::Strict,
    };

    // Solve it
    let records = resolvo::Solver.solve(task).into_diagnostic()?;

    Ok(records)
}

/// Get networking Client and fetch [`SparseRepoData`] for the given channels and
/// current platform using the client
///
/// # Returns
///
/// The network client and the fetched sparse repodata
pub(super) async fn get_client_and_sparse_repodata(
    channels: impl IntoIterator<Item = &'_ Channel>,
    platform: Platform,
    config: &Config,
) -> miette::Result<(
    ClientWithMiddleware,
    IndexMap<(Channel, Platform), SparseRepoData>,
)> {
    let authenticated_client = build_reqwest_clients(Some(config)).1;
    let platform_sparse_repodata =
        repodata::fetch_sparse_repodata(channels, [platform], &authenticated_client, Some(config))
            .await?;
    Ok((authenticated_client, platform_sparse_repodata))
}

/// Find the globally installed package with the given [`PackageName`]
///
/// # Returns
///
/// The PrefixRecord of the installed package
pub(super) async fn find_installed_package(
    package_name: &PackageName,
) -> miette::Result<PrefixRecord> {
    let BinEnvDir(bin_prefix) = BinEnvDir::from_existing(package_name).await.or_else(|_| {
        miette::bail!(
            "Package {} is not globally installed",
            package_name.as_source()
        )
    })?;
    let prefix = Prefix::new(bin_prefix);
    find_designated_package(&prefix, package_name).await
}

/// Find the designated package in the given [`Prefix`]
///
/// # Returns
///
/// The PrefixRecord of the designated package
pub async fn find_designated_package(
    prefix: &Prefix,
    package_name: &PackageName,
) -> miette::Result<PrefixRecord> {
    let prefix_records = prefix.find_installed_packages(None).await?;
    prefix_records
        .into_iter()
        .find(|r| r.repodata_record.package_record.name == *package_name)
        .ok_or_else(|| miette::miette!("could not find {} in prefix", package_name.as_source()))
}
