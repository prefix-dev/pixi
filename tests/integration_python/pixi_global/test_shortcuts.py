from pathlib import Path
import tomllib
from typing import List

import tomli_w
from ..common import verify_cli_command, ExitCode, CURRENT_PLATFORM
from abc import ABC, abstractmethod
import pytest
from dataclasses import dataclass


@dataclass
class SetupData:
    pixi_home: Path
    data_home: Path
    env: dict[str, str]


@pytest.fixture
def setup_data(tmp_path: Path) -> SetupData:
    pixi_home = tmp_path / "pixi_home"
    data_home = tmp_path / "data_home"
    env = {
        "PIXI_HOME": str(pixi_home),
        "HOME": str(data_home),  # Used for macOS and Linux
        "MENUINST_FAKE_DIRECTORIES": str(data_home),  # Used for Windows
    }
    return SetupData(pixi_home=pixi_home, data_home=data_home, env=env)


class PlatformConfig(ABC):
    @abstractmethod
    def _shortcut_paths(self, data_home: Path, name: str) -> List[Path]:
        pass

    @abstractmethod
    def shortcut_exists(self, data_home: Path, name: str) -> bool:
        """Given the name of a shortcut, return whether it exists or not."""
        pass


class LinuxConfig(PlatformConfig):
    def _shortcut_paths(self, data_home: Path, name: str) -> List[Path]:
        return [data_home / ".local" / "share" / "applications" / f"{name}_{name}.desktop"]

    def shortcut_exists(self, data_home: Path, name: str) -> bool:
        return self._shortcut_paths(data_home, name).pop().is_file()


class MacOSConfig(PlatformConfig):
    def _shortcut_paths(self, data_home: Path, name: str) -> List[Path]:
        return [data_home / "Applications" / f"{name}.app"]

    def shortcut_exists(self, data_home: Path, name: str) -> bool:
        return self._shortcut_paths(data_home, name).pop().is_dir()


class WindowsConfig(PlatformConfig):
    def _shortcut_paths(self, data_home: Path, name: str) -> List[Path]:
        return [data_home / "Desktop" / f"{name}.lnk"]

    def shortcut_exists(self, data_home: Path, name: str) -> bool:
        for path in self._shortcut_paths(data_home, name):
            print(path)
            if not path.is_file():
                return False
        return True


def get_platform_config(platform: str) -> PlatformConfig:
    if platform == "linux-64":
        return LinuxConfig()
    elif platform in {"osx-arm64", "osx64"}:
        return MacOSConfig()
    elif platform == "win-64":
        return WindowsConfig()
    else:
        raise ValueError(f"Unsupported platform: {platform}")


def verify_shortcuts_exist(
    data_home: Path,
    shortcut_names: list[str],
    expected_exists: bool,
) -> None:
    """Verify if the specified shortcuts exist or not on the given system."""
    # Using the key to get the platform-specific configuration, to force a KeyError if the key is not found
    system = CURRENT_PLATFORM
    config = get_platform_config(system)
    for name in shortcut_names:
        assert config.shortcut_exists(data_home, name) == expected_exists, (
            f"Shortcut '{name}' {'should' if expected_exists else 'should not'} exist on {system}"
        )


def test_sync_creation_and_removal(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    """Test shortcut creation and removal with sync."""

    # Setup manifest with given shortcuts
    manifests = setup_data.pixi_home.joinpath("manifests")
    manifests.mkdir(parents=True)
    manifest = manifests.joinpath("pixi-global.toml")
    toml = f"""
    [envs.test]
    channels = ["{shortcuts_channel_1}"]
    dependencies = {{ pixi-editor = "*" }}
    """
    manifest.write_text(toml)

    # Verify no shortcuts exist after sync
    verify_cli_command([pixi, "global", "sync"], ExitCode.SUCCESS, env=setup_data.env)
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=False)

    parsed_toml = tomllib.loads(toml)
    parsed_toml["envs"]["test"]["shortcuts"] = ["pixi-editor"]
    manifest.write_text(tomli_w.dumps(parsed_toml))

    # # Run sync and verify
    verify_cli_command([pixi, "global", "sync"], ExitCode.SUCCESS, env=setup_data.env)
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=True)

    # test removal of shortcuts
    del parsed_toml["envs"]["test"]["shortcuts"]
    manifest.write_text(tomli_w.dumps(parsed_toml))
    verify_cli_command([pixi, "global", "sync"], ExitCode.SUCCESS, env=setup_data.env)
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=False)


def test_sync_empty_shortcut_list(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    # Setup manifest with given shortcuts
    manifests = setup_data.pixi_home.joinpath("manifests")
    manifests.mkdir(parents=True)
    manifest = manifests.joinpath("pixi-global.toml")
    toml = f"""
    [envs.test]
    channels = ["{shortcuts_channel_1}"]
    dependencies = {{ pixi-editor = "*" }}
    shortcuts = ["pixi-editor"]
    """
    manifest.write_text(toml)

    # Run sync and verify
    verify_cli_command([pixi, "global", "sync"], ExitCode.SUCCESS, env=setup_data.env)
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=True)

    # Set shortcuts to empty list
    parsed_toml = tomllib.loads(toml)
    parsed_toml["envs"]["test"]["shortcuts"] = []
    manifest.write_text(tomli_w.dumps(parsed_toml))
    verify_cli_command([pixi, "global", "sync"], ExitCode.SUCCESS, env=setup_data.env)
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=False)


def test_sync_removing_environment(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    # Setup manifest with given shortcuts
    manifests = setup_data.pixi_home.joinpath("manifests")
    manifests.mkdir(parents=True)
    manifest = manifests.joinpath("pixi-global.toml")
    toml = f"""
    [envs.test]
    channels = ["{shortcuts_channel_1}"]
    dependencies = {{ pixi-editor = "*" }}
    shortcuts = ["pixi-editor"]
    """
    manifest.write_text(toml)

    # Run sync and verify
    verify_cli_command([pixi, "global", "sync"], ExitCode.SUCCESS, env=setup_data.env)
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=True)

    # Remove environment
    parsed_toml = tomllib.loads(toml)
    del parsed_toml["envs"]["test"]
    manifest.write_text(tomli_w.dumps(parsed_toml))
    verify_cli_command([pixi, "global", "sync"], ExitCode.SUCCESS, env=setup_data.env)
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=False)


def test_sync_duplicate_shortcuts(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    """Test shortcut creation and removal with sync."""

    # Setup manifest with given shortcuts
    manifests = setup_data.pixi_home.joinpath("manifests")
    manifests.mkdir(parents=True)
    manifest = manifests.joinpath("pixi-global.toml")
    toml = f"""
    [envs.test1]
    channels = ["{shortcuts_channel_1}"]
    dependencies = {{ pixi-editor = "*" }}
    shortcuts = ["pixi-editor"]

    [envs.test2]
    channels = ["{shortcuts_channel_1}"]
    dependencies = {{ pixi-editor = "*" }}
    shortcuts = ["pixi-editor"]
    """
    manifest.write_text(toml)

    # Verify no shortcuts exist after sync
    verify_cli_command(
        [pixi, "global", "sync"],
        ExitCode.FAILURE,
        env=setup_data.env,
        stderr_contains="Duplicated shortcut names found: pixi-editor",
    )


def test_sync_unavailable_shortcuts(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    """Test shortcut creation and removal with sync."""

    # Setup manifest with given shortcuts
    manifests = setup_data.pixi_home.joinpath("manifests")
    manifests.mkdir(parents=True)
    manifest = manifests.joinpath("pixi-global.toml")
    toml = f"""
    [envs.test1]
    channels = ["{shortcuts_channel_1}"]
    dependencies = {{ pixi-editor = "*" }}
    shortcuts = ["unavailable-shortcut"]
    """
    manifest.write_text(toml)

    # Verify no shortcuts exist after sync
    verify_cli_command(
        [pixi, "global", "sync"],
        ExitCode.FAILURE,
        env=setup_data.env,
        stderr_contains="the following shortcuts are requested but not available: unavailable-shortcut",
    )


def test_install_simple(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    """Test shortcut creation with install."""

    # Verify no shortcuts exist after sync
    verify_cli_command(
        [pixi, "global", "install", "--channel", shortcuts_channel_1, "pixi-editor"],
        ExitCode.SUCCESS,
        env=setup_data.env,
    )

    # Verify manifest
    manifest = setup_data.pixi_home.joinpath("manifests", "pixi-global.toml")
    parsed_toml = tomllib.loads(manifest.read_text())
    assert parsed_toml["envs"]["pixi-editor"]["shortcuts"] == ["pixi-editor"]

    # Verify shortcut exists
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=True)


def test_install_no_shortcut(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    """Test shortcut creation with install where `--no-shortcut` was passed."""

    # Verify no shortcuts exist after sync
    verify_cli_command(
        [
            pixi,
            "global",
            "install",
            "--channel",
            shortcuts_channel_1,
            "--no-shortcut",
            "pixi-editor",
        ],
        ExitCode.SUCCESS,
        env=setup_data.env,
    )

    # Verify manifest
    manifest = setup_data.pixi_home.joinpath("manifests", "pixi-global.toml")
    parsed_toml = tomllib.loads(manifest.read_text())
    assert "shortcuts" not in parsed_toml["envs"]["pixi-editor"]

    # Verify shortcut does not exist
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=False)


def test_remove_shortcut(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    # Verify no shortcuts exist after sync
    verify_cli_command(
        [pixi, "global", "install", "--channel", shortcuts_channel_1, "pixi-editor"],
        ExitCode.SUCCESS,
        env=setup_data.env,
    )

    # Verify manifest
    manifest = setup_data.pixi_home.joinpath("manifests", "pixi-global.toml")
    parsed_toml = tomllib.loads(manifest.read_text())
    assert parsed_toml["envs"]["pixi-editor"]["shortcuts"] == ["pixi-editor"]

    # Verify shortcut exists
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=True)

    # Remove shortcut
    verify_cli_command(
        [pixi, "global", "shortcut", "remove", "pixi-editor"],
        ExitCode.SUCCESS,
        env=setup_data.env,
    )

    # Verify removal from manifest
    parsed_toml = tomllib.loads(manifest.read_text())
    assert parsed_toml["envs"]["pixi-editor"]["shortcuts"] != ["pixi-editor"]

    # Verify shortcut does not exist
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=False)


def test_add_shortcut(
    pixi: Path,
    setup_data: SetupData,
    shortcuts_channel_1: str,
) -> None:
    verify_cli_command(
        [
            pixi,
            "global",
            "install",
            "--no-shortcuts",
            "--channel",
            shortcuts_channel_1,
            "pixi-editor",
        ],
        ExitCode.SUCCESS,
        env=setup_data.env,
    )

    # Verify manifest
    manifest = setup_data.pixi_home.joinpath("manifests", "pixi-global.toml")
    parsed_toml = tomllib.loads(manifest.read_text())
    assert parsed_toml["envs"]["pixi-editor"].get("shortcuts") is None

    # Verify shortcut exists
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=False)

    # Add shortcut
    verify_cli_command(
        [pixi, "global", "shortcut", "add", "pixi-editor", "--environment", "pixi-editor"],
        ExitCode.SUCCESS,
        env=setup_data.env,
    )

    # Verify addition to manifest
    parsed_toml = tomllib.loads(manifest.read_text())
    assert parsed_toml["envs"]["pixi-editor"]["shortcuts"] == ["pixi-editor"]

    # Verify shortcut exists
    verify_shortcuts_exist(setup_data.data_home, ["pixi-editor"], expected_exists=True)
