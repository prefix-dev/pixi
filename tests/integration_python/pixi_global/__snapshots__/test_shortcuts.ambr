# serializer version: 1
# name: test_sync_shortcuts[Linux-desktop-file]
  '''
  [Desktop Entry]
  Type=Application
  Encoding=UTF-8
  Name=pixi-editor
  Exec=bash -c 'gedit %F'
  Terminal=false
  Icon=/var/home/julian/Projekte/github.com/prefix-dev/pixi-1/pytest-temp/test_sync_shortcuts0/pixi_home/envs/test/Menu/pixi-icon.png
  Comment=Scientific Python Development Environment
  Categories=Development;Science;
  MimeType=text/x-pixi;
  
  '''
# ---
